p2p/net/conn/transport.go

```diff
@@ -0,0 +1,191 @@
+package conn
+
+import (
+	"net"
+
+	ma "github.com/ipfs/go-ipfs/Godeps/_workspace/src/github.com/jbenet/go-multiaddr"
+	manet "github.com/ipfs/go-ipfs/Godeps/_workspace/src/github.com/jbenet/go-multiaddr-net"
+	mautp "github.com/ipfs/go-ipfs/Godeps/_workspace/src/github.com/jbenet/go-multiaddr-net/utp"
+	reuseport "github.com/ipfs/go-ipfs/Godeps/_workspace/src/github.com/jbenet/go-reuseport"
+)
+
+type Transport interface {
+	manet.Listener
+	ProtoDialer
+}
+
+type ProtoDialer interface {
+	Dial(raddr ma.Multiaddr) (manet.Conn, error)
+	Matches(ma.Multiaddr) bool
+}
+
+type TcpReuseTransport struct {
+	list  manet.Listener
+	laddr ma.Multiaddr
+
+	rd       reuseport.Dialer
+	madialer manet.Dialer
+}
+
+var _ Transport = (*TcpReuseTransport)(nil)
+
+func NewTcpReuseTransport(base manet.Dialer, laddr ma.Multiaddr) (*TcpReuseTransport, error) {
+	rd := reuseport.Dialer{base.Dialer}
+
+	list, err := manet.Listen(laddr)
+	if err != nil {
+		return nil, err
+	}
+
+	// get the local net.Addr manually
+	la, err := manet.ToNetAddr(laddr)
```

Comment:
Also, i do not think the transport should start listening just by being created. `reuseport` should allow listening _after_, (yes after a dial has opened on the same port).